e2e/app/element/base-element.ts

```diff
@@ -7,69 +8,69 @@ import {ClickOptions, ElementHandle, Page, WaitForSelectorOptions} from 'puppete
 
   static asBaseElement(page: Page, elem: ElementHandle): BaseElement {
-    return new BaseElement(page, elem);
+    const baseElement = new BaseElement({puppeteerPage: page});
+    baseElement.element = elem;
+    return baseElement;
   }
 
-  protected readonly page: Page;
-  protected css: string;
-  protected xpath: string;
-  protected element: ElementHandle;
+  protected _element: ElementHandle;
+  protected readonly _xpath;
+  protected readonly _page: Page;
+  protected readonly _container: Container;
 
-  constructor(aPage: Page, aElement?: ElementHandle) {
-    this.page = aPage;
-    this.element = aElement || undefined;
+  constructor(pageOptions: {puppeteerPage: Page, container?: Container},
+              selector: {xpath?: string, testId?: string} = {}) {
+    this._page = pageOptions.puppeteerPage;
+    this._container = pageOptions.container || undefined;
+    if (selector.testId !== undefined) {
+      this._xpath = `//*[@data-test-id="${selector.testId}"]`;
+    } else if (selector.xpath !== undefined) {
+      this._xpath = (this._container === undefined) ? selector.xpath : pageOptions.container.xpath + selector.xpath;
+    }
   }
 
-  /**
-   * Find element with wait.
-   * @param cssSelector
-   * @param waitOptions
-   */
-  async withCss(cssSelector: string, waitOptions?: WaitForSelectorOptions): Promise<ElementHandle> {
-    this.css = cssSelector;
-    this.element = await this.page.waitForSelector(this.css, waitOptions);
-    return this.element;
+  set element(handle: ElementHandle) {
+    this._element = handle;
   }
 
-  /**
-   * Find element with wait.
-   * @param xpathSelector
-   * @param waitOptions
-   */
-  async withXpath(xpathSelector: string, waitOptions?: WaitForSelectorOptions): Promise<ElementHandle> {
-    this.xpath = xpathSelector;
-    this.element = await this.page.waitForXPath(this.xpath, waitOptions);
-    return this.element;
+  get element() {
+    return this._element;
   }
 
-   /**
-    * Find first element without wait for.
-    */
-  async findByCss(cssSelector: string,): Promise<ElementHandle | null> {
-    this.css = cssSelector;
-    this.element = await this.page.$(this.css);
-    return this.element;
+  get xpath() {
+    return this._xpath;
   }
 
-   /**
-    * Find first element without wait for.
-    */
-  async findByXpath(xpathSelector: string): Promise<ElementHandle | null> {
-    this.xpath = xpathSelector;
-    const found = await this.page.$x(this.xpath);
-    if (found.length > 0) {
-      this.element = found[0];
+  get container() {
+    return this._container;
+  }
+
+  async findElement(waitOptions: WaitForSelectorOptions = { visible: true }): Promise<ElementHandle | null> {
+    if (this._element !== undefined) {
+      return this._element;
+    }
+    if (this._xpath !== undefined) {
+      this._element = await this._page.waitForXPath(this._xpath, waitOptions);
     } else {
-      this.element = null;
+      this._element = null;
```

Comment:
So now we're introducing `null` in addition to the convention of using `undefined`. It's going to be easy to write bugs by confusing these, considering how many places are explicitly checking for `undefined`.

Also, please document the lazy evaluation here. I think it's safe, though it could affect timings potentially if used in a tight loop.

Reply:
I can buy that.